internal/handler/report_api_csp.go

```diff
@@ -0,0 +1,183 @@
+			"original_policy":     violation.Body.OriginalPolicy,
+			"disposition":         violation.Body.Disposition,
+			"status_code":         violation.Body.StatusCode,
+			"source_file":         violation.Body.SourceFile,
+			"line_number":         violation.Body.LineNumber,
+			"column_number":       violation.Body.ColumnNumber,
+			"metadata":            metadata,
+			"path":                r.URL.Path,
+		}
+
+		if vrh.TruncateQueryStringFragment {
+			lf["document_uri"] = utils.TruncateQueryStringFragment(violation.Body.DocumentURL)
+			lf["referrer"] = utils.TruncateQueryStringFragment(violation.Body.Referrer)
+			lf["blocked_uri"] = utils.TruncateQueryStringFragment(violation.Body.BlockedURL)
+			lf["source_file"] = utils.TruncateQueryStringFragment(violation.Body.SourceFile)
+		}
+
+		if vrh.LogClientIP {
+			ip, err := utils.GetClientIP(r)
+			if err != nil {
+				vrh.Logger.Warnf("unable to parse client ip: %s", err)
+			}
+			lf["client_ip"] = ip.String()
+		}
+
+		if vrh.LogTruncatedClientIP {
+			ip, err := utils.GetClientIP(r)
+			if err != nil {
+				vrh.Logger.Warnf("unable to parse client ip: %s", err)
+			}
+			lf["client_ip"] = utils.TruncateClientIP(ip)
+		}
+
+		vrh.Logger.WithFields(lf).Info()
+	}
+
+}
+
+func (vrh *ReportAPIViolationReportHandler) validateViolation(r ReportAPIReports) error {
+	for _, violation := range r.Reports {
+		if violation.Type != "csp-violation" {
+			continue // Skip the rest of the loop and move to the next iteration
+		}
+		for _, value := range vrh.BlockedURIs {
+			if strings.HasPrefix(violation.Body.BlockedURL, value) {
+				err := fmt.Errorf("blocked URI ('%s') is an invalid resource", value)
+				return err
+			}
+		}
+		if !strings.HasPrefix(violation.Body.DocumentURL, "http") {
+			return fmt.Errorf("document URI ('%s') is invalid", violation.Body.DocumentURL)
+		}
+	}
+
+	return nil
+}
+
+func ReportAPICorsHandler(w http.ResponseWriter, r *http.Request) {
+	origin := r.Header.Get("Origin")
+	method := r.Header.Get("Access-Control-Request-Method")
+	header := r.Header.Get("Access-Control-Request-Headers")
+	allow_origin := utils.Ternary(origin != "", origin, "*")
+	allow_method := utils.Ternary(method != "", method, "*")
+	allow_header := utils.Ternary(header != "", header, "*")
+	// Special handling due to bug in Chrome
+	// https://bugs.chromium.org/p/chromium/issues/detail?id=1152867
+	w.Header().Set("Access-Control-Allow-Origin", allow_origin)
+	w.Header().Set("Access-Control-Allow-Methods", allow_method)
+	w.Header().Set("Access-Control-Max-Age", "60")
+	w.Header().Set("Access-Control-Allow-Headers", allow_header)
+	w.Header().Set("vary", "Origin, Access-Control-Request-Method, Access-Control-Request-Headers")
+
+	w.Header().Set("Cross-Origin-Resource-Policy", "cross-origin")
+	w.Header().Set("Content-Type", "text/plain;charset=UTF-8")
+	w.Header().Set("Server", "cloudflare")
+	w.WriteHeader(http.StatusOK)
+	w.Write([]byte("OK"))
```

Comment:
do we need to send a response here or is a HTTP 200/202 ok? i'm not seeing anything in the HTTP spec requiring this so we should just be able to send the status and bail.

Reply:
@jacobbednarz in my testing the reporting api implementation on the browser side was picky. I wanted to make the responses resemble report-uri.com as closely as possible.
having that in mind - if no content is sent a 204 would be more appropriate and that caused some issues for me during testing.

Reply:
that's fine if it's problematic. we can leave it as is.